importccl: cleanup how we ignore stmts in IMPORT PGDUMP

adityamaru opened the pull request:
There are two kinds of unsupported statements when processing an IMPORT
PGDUMP:
- Unparseable
- Parseable, but unsupported

Previously, we had a set of regex statements to skip over statements
which fell into the first category. Those which fell into the second
category were skipped when we saw an AST Node of that kind.

This change does not ignore any additional statements, but removes the
regex. It does this by leaning on existing yacc grammar rules, or adding
new "unsupported" rules. These unsupported rules raise a special kind of
error which we catch during IMPORT PGDUMP, and is decorated with
required information.

This change also introduces a `ignore_unsupported` flag to IMPORT PGDUMP
which will allow users to skip all the stmts (of both categories
mentioned above). If not set, we will fail with either a parse or
unsupported stmt error.

Release note (sql change): New IMPORT PGDUMP option `ignore_unsupported`
to skip over all the unsupported PGDUMP stmts. The collection of these
statements will be appropriately documented.

adityamaru committed:
importccl: cleanup how we ignore stmts in IMPORT PGDUMP

There are two kinds of unsupported statements when processing an IMPORT
PGDUMP:
- Unparseable
- Parseable, but unsupported

Previously, we had a set of regex statements to skip over statements
which fell into the first category. Those which fell into the second
category were skipped when we saw an AST Node of that kind.

This commit is responsible for the following:

- Removes regex and adds/modifies yacc rules to encompass these regex
  representations.

- Introduces a special UnsupportedNode which backs all yacc rules
  which previously returned an unimplemented error.

- Introduces an IMPORT PGDUMP option to `ignore_unsupported` statements.

- All statements which are not parseable will now be ignored if the
  option is set.

Note: the eventual goal is to have the `ignore_unsupported` flag also
allow users to ignore parseable, but unsupported statements. This will
be done in the next commit.

Release note (sql change): New IMPORT PGDUMP option `ignore_unsupported`
to skip over all the unsupported PGDUMP stmts. The collection of these
statements will be appropriately documented.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57827)
<!-- Reviewable:end -->

adityamaru commented:
@dt we discussed that maintaining existing semantics of always ignoring the previously regex-ed statements is less controversial, but while implementing I started liking the idea of biting the bullet once and grouping all ignored statements under the new option. Having this mixed state where some features are ignored by default and others guarded by this option seems a little messy. In my opinion, if the user wants to ignore **all unsupported stmts** they should supply that option going forward. Open to discussion about this, the change to go either way is simple!

Also open to suggestions on how to convey to the user what we have skipped. I was thinking write to `nodelocal` of the coordinator node if the user has given us externalio access?

pbardea commented on pkg/ccl/importccl/import_stmt_test.go:
```diff
@@ -5417,6 +5394,51 @@ func TestImportPgDump(t *testing.T) {
 	})
 }
 
+func TestImportPgDumpIgnoredStmts(t *testing.T) {
```
Should we also add tests for unimplemented but "supported" statements to document their interaction with the `ignore_unsupported` flag?

adityamaru replied:
Yep, I will add this as part of the second commit which will be ignoring:
- Statements which we silently ignored previously.
- Statements which we error'ed on before during the tree.Node evaluation, but now ignore if the setting is set.

pbardea commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -99,6 +99,8 @@ const (
 	avroSchema    = "schema"
 	avroSchemaURI = "schema_uri"
 
+	pgDumpIgnoreAllUnsupported = "ignore_unsupported"
```
I think as a user I would find it confusing that there are statements which still return some sort of error like "unimplemented" with this flag (please correct me if I'm reading this wrong, but a statement `CREATE OPERATOR` is not "unsupported"). As you mentioned in the commit message, these would be documented but I still think it might be confusing?

Does it make sense to ignore all unimplemented nodes if we can write them to a rejected file?

adityamaru replied:
I'm a little confused, I'm going to try and categorize the types of errors we can see, and then we can decide what should be guarded by this flag.

1. There are statements like `CREATE OPERATOR` which return an unimplemented error on parsing. There are several of these in sql.y. This PR replaced some (previously caught by our regex) of those unimplemented errors to instead be backed by an unsupported node, so that when we see such a node on parse.Parse() we can skip over it.

I guess my first question is: do we want to skip **all** sql statements which are unimplemented in our parser, or do we want to have control and only skip a subset of those. This would mean for the rest, the user would see an unable to parse error.

If the answer to the above question is yes, I would need to wholesale change all the sql.y grammar rules to be backed by our unsupported node. Albeit this could be done by just changing the `unimplemented()` methods they currently call.

2. The second group of statements is the ones that have associated tree.Nodes. A statement with a tree.Node is either "supported" by IMPORT, "unsupported and ignored", or just "unsupported" and in the last case, we will throw an error. Here I think I agree with you that if we see something which made it past parsing, but we don't support it then it should not throw an error. This PR doesn't do that yet.

Let me know what you think or if I'm misunderstanding your comment. Going to cc: @mwang1026 in case he has thoughts too.

mwang1026 replied:
I agree it's confusing UX for a user that said "I want to ignore things that are unsupported" for it to ignore just a subset, especially if we have a "reject" file

pbardea replied:
> I guess my first question is: do we want to skip all sql statements which are unimplemented in our parser, or do we want to have control and only skip a subset of those.

I think that that's also what I was after. I think that moving the ignoring logic from regexes to sql.y is good and can be done on it's own. My concern was that before we were silently ignoring these, so it wasn't surprising to a user that they would run an import and reach some unsupported statement. However, now if they want these statements to be unsupported they have to run the import with the `ignore_unsupported`, and they might _still_ get "unsupported" errors which is confusing.

I don't think I have a good enough grasp of why some nodes are "unsupported and ignored" or just "unsupported" to have a more concrete opinion on the best way to resolve this though.

adityamaru replied:
@pbardea just to re-outline the new approach since there has been a lot of discussion:

The `ignore_unsupported` option will skip the following genres of statements:
- All statements that have rules in sql.y and return an `unimplemented` error.
- All statements that were previously ignored silently egs: `BEGIN`, `COMMIT` etc
- All statements that were previously successfully parsed, but returned an `unsupported` error during tree.Node evaluation.

The first commit only addresses the first bullet point. Subsequent commits will tackle the next two bullet points.

adityamaru commented on pkg/ccl/importccl/import_stmt_test.go:
```diff
@@ -5417,6 +5394,51 @@ func TestImportPgDump(t *testing.T) {
+	}))
+	defer srv.Close()
+	t.Run("ignore-unsupported", func(t *testing.T) {
+		sqlDB.Exec(t, "IMPORT PGDUMP ($1) WITH ignore_unsupported", srv.URL)
```
added two statements which should not be ignored even when the flag is set. Now that every unparseable statement in sql.y can technically be ignored, I'm not sure what a good test would look like. I just picked up these statements from an older test which relied on the regex.

adityamaru commented on pkg/sql/sem/tree/unsupported.go:
```diff
@@ -0,0 +1,26 @@
+// an import.
+type Unsupported struct {
+	Err                    error
+	SkipDuringImportPGDump bool
```
removed the field. I've kept the name as Unsupported since it now backs all unparseable SQL statements. We do only use this fact during an import but I went with a more generic name. Could go either way though.

adityamaru commented on pkg/sql/parser/sql.y:
```diff
@@ -3894,6 +3921,10 @@ revoke_stmt:
   }
+| REVOKE privileges ON SEQUENCE error
+  {
+    return unimplementedWithUnsupportedNode(sqllex, "REVOKE ON SEQUENCE", true)
```
removed the field.

adityamaru commented on pkg/ccl/importccl/read_import_pgdump.go:
```diff
@@ -535,13 +533,22 @@ func readPostgresStmt(
 	case *tree.Analyze:
-		// ANALYZE is syntatictic sugar for CreateStatistics. It can be ignored because
-		// the auto stats stuff will pick up the changes and run if needed.
+	// ANALYZE is syntatictic sugar for CreateStatistics. It can be ignored because
```
done.

pbardea reviewed (approved):
LGTM

adityamaru committed:
importccl: guard all silently ignored PGDUMP stmts with option

Previously, we would silently ignore several statements we might come
across in a PGDUMP file. These statements can be parsed by CRDB but are
not implemented by IMPORT.

This change brings most of these silently ignored stmts under the
umbrella of the newly introduced `ignore_unsupported` option. The
motivation for this change is to further improve the UX of this option.
If specified, we will ignore unsupported stmts, else we will fail the
IMPORT.

This will result in IMPORTs that previously succeeded by silently
ignoring some stmts to fail. The solution is to add the
`ignore_unsupported` option.

Release note (sql change): Users will now need to specify
`ignore_unsupported` to ignore all unsupported import stmts during an
IMPORT PGDUMP.

miretskiy reviewed (changes requested):
> Reviewed 2 of 10 files at r1, 6 of 8 files at r2, 5 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57827)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @pbardea)

___
*[out, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST0NE-Cfym12H2vr7S:-MST0NE-Cfym12H2vr7T:b9inhl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/out#L1)):*
> ```
> Running make with -j16
> ```

file Accidentally included?
___
*[output, line 0 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST0SJkB3kEmVYeveG0:-MST0SJkB3kEmVYeveG1:b-faicfr) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/output#L0)):*
File accidentally included?
___
*[pkg/ccl/importccl/import_processor_test.go, line 957 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST0e2o2fl7R4ysKZ1O:-MST0e2o2fl7R4ysKZ1P:b-50rwj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/import_processor_test.go#L957)):*
> ```Go
> 		PgDump: roachpb.PgDumpOptions{
> 			MaxRowSize:        64 * 1024,
> 			IgnoreUnsupported: true,
> ```

Should you update some files in  pkg/ccl/importccl/testdata/pgdump/ to add some unsupported cruft?
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST1IVr43He23JY8h3e:-MST1IVr43He23JY8h3f:bwg2wcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L108)):*
> ```Go
> 			// skipped.
> 			if errors.HasType(err, (*tree.Unsupported)(nil)) {
> 				if p.ignoreUnsupportedStmts {
> ```

combine these two ifs? I would check for if p.ignoreUnsupported first.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 154 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST1QzMAcyh_H3X35kK:-MST1QzMAcyh_H3X35kL:b-hmxie2) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L154)):*
> ```Go
> 
> var (
> 	ignoreComments = regexp.MustCompile(`^\s*(--.*)`)
> ```

i'm curious -- can't our lexer skip comments for us?
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 398 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST1WQQ4AyWhMTkX3qd:-MST1WQQ4AyWhMTkX3qe:byf421e) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L398)):*
> ```Go
> 				if cmd.IfNotExists {
> 					if ignoreUnsupportedStmts {
> 						// Write to shunt.
> ```

is that a TODO?
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 426 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST2-OCCUtdHC8nhs9b:-MST2-OCCUtdHC8nhs9c:bxfbf3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L426)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> 			case *tree.AlterTableValidateConstraint:
> 				if ignoreUnsupportedStmts {
> 					// Write to shunt.
> 					continue
> 				}
> 				return errors.Errorf("unsupported statement: %s", stmt)
> ```
</details>

Should this branch be dropped altogether? Same as default...
___
*[pkg/sql/parser/sql.y, line 2718 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST07f157FD_bDeRDgr:-MST07f157FD_bDeRDgs:b-7expk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/sql/parser/sql.y#L2718)):*
> ```Yacc
>   }
> 
> alter_unsupported_stmt:
> ```

I think you need to add appropriate parse tests (pkg/sql/parse_test.go) for changes in sql.y
___
*[pkg/sql/sem/tree/unsupported.go, line 13 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST-xHm8NvZoxxAuucj:-MST-xHm8NvZoxxAuuck:b-u0oe38) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/sql/sem/tree/unsupported.go#L13)):*
> ```Go
> package tree
> 
> var _ error = &Unsupported{}
> ```

should this be called UnsupportedError?


<!-- Sent from Reviewable.io -->

adityamaru committed:
importccl: add option to log unsupported stmts

This is the last commit, that adds an `ignored_stmt_log` options to
IMPORT PGDUMP. This option specifies the destination we will log the
statements that we skip over during an import. This option can only be
used in conjunction with `ignore_unsupported`, else the IMPORT will
fail.

Currently, we will write to two files during the import. One during the
schema parsing phase, and another during the data ingestion phase. The
files will be called:

`unsupported-data-stmts`: Contains unparseable stmts, and unsupported
DML stmts.

`unsupported-schema-stmts`: Contains unparseable stmts, and unsupported
DDL stmts.

Release note (sql change): New IMPORT PGDUMP option `ignored_stmt_log`
that allows users to specify where they would like to log stmts that
have been skipped during an import, by virtue of being unsupported.

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57827)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[out, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST0NE-Cfym12H2vr7S:-MSUMg6z6H4a8X0moPZf:bx2h8cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/out#L1)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

file Accidentally included?
</blockquote></details>

ah sorry, deleted.
___
*[output, line  at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST0SJkB3kEmVYeveG0:-MSUMiP-7svX19F4HpND:b-k4nm9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/output#L)):*
> ah sorry, deleted.
deleted oops.
___
*[pkg/ccl/importccl/import_processor_test.go, line 957 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST0e2o2fl7R4ysKZ1O:-MSUgjy_7lWubaZKSPNF:b-hbawh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/import_processor_test.go#L957)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Should you update some files in  pkg/ccl/importccl/testdata/pgdump/ to add some unsupported cruft?
</blockquote></details>

some of the files already have cruft that was silently being ignored before.

egs: simple.sql has 
`SELECT pg_catalog.set_config('search_path', '', false);`
`ALTER TABLE public.simple OWNER TO postgres;`

db.sql has:
`CREATE EXTENSION IF NOT EXISTS plpgsql WITH SCHEMA pg_catalog;`
`COMMENT ON EXTENSION plpgsql IS 'PL/pgSQL procedural language';`

I added a few more statements to simple.sql.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST1IVr43He23JY8h3e:-MSUNV7sCd18ZO2cj_H3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L108)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

combine these two ifs? I would check for if p.ignoreUnsupported first.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 154 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST1QzMAcyh_H3X35kK:-MSUfwXaBHC-i2CC4ikz:b-g28fwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L154)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

i'm curious -- can't our lexer skip comments for us?
</blockquote></details>

haven't had a chance to look into this yet, but I will before merging this PR.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 398 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST1WQQ4AyWhMTkX3qd:-MSUNW3hEzcHqu2n7t8E:b-gp8cyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L398)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

is that a TODO?
</blockquote></details>

yup, i'll replace these in the next commit.
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 426 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST2-OCCUtdHC8nhs9b:-MSUNiR6AJdTTB0kioa1:b-o6g8tu) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/ccl/importccl/read_import_pgdump.go#L426)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 			case *tree.AlterTableValidateConstraint:
> 				if ignoreUnsupportedStmts {
> 					// Write to shunt.
> 					continue
> 				}
> 				return errors.Errorf("unsupported statement: %s", stmt)
> ```
Should this branch be dropped altogether? Same as default...
</blockquote></details>

good point, deleted!
___
*[pkg/sql/parser/sql.y, line 2718 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST07f157FD_bDeRDgr:-MSUcooS49g4yDZk_TIp:b-p4skgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/sql/parser/sql.y#L2718)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think you need to add appropriate parse tests (pkg/sql/parse_test.go) for changes in sql.y
</blockquote></details>

yup, missed that, added!
___
*[pkg/sql/sem/tree/unsupported.go, line 13 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MST-xHm8NvZoxxAuucj:-MSUfY-AB7-MiUIKQ87j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f169e969426672ff886a3fd6632ad19424c4be/pkg/sql/sem/tree/unsupported.go#L13)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should this be called UnsupportedError?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 4 of 8 files at r4, 7 of 10 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57827)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/importccl/import_stmt.go, line 1281 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MS_RV1w-rXTNUowbQpa:-MS_RV1w-rXTNUowbQpb:b-tgyvf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c645c03431f34adaec14b93dcc4a0ad2919aed5/pkg/ccl/importccl/import_stmt.go#L1281)):*
> ```Go
> 
> 	// logBuffer holds the string to be flushed to the ignoreUnsupportedLogDest.
> 	logBuffer *bytes.Buffer
> ```

a bit concerned we could blow up server if we try to import pgdump file w/ slew of unsupported statements...

I would be okay just limiting the number of errors we emit; followed by message... "too many errors -- rest skipped"...
Or.. i don't know, maybe just collect very small number of those errored statements (say 5-10) and just send a notice at the end with the
"sample" of errors.

Alternatively, maybe just drop logging to external storage?


<!-- Sent from Reviewable.io -->

adityamaru commented:
This is RFAL! I added a count and a limit to the number of lines we log.

pbardea commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -99,6 +100,12 @@ const (
 
+	pgDumpIgnoreAllUnsupported     = "ignore_unsupported"
+	pgDumpIgnoreShuntFileDest      = "ignored_stmt_log"
+	pgDumpUnsupportedSchemaStmtLog = "unsupported-schema-stmts"
```
Let's change these separators from `-` to `_` for consistency with the other ones.

adityamaru replied:
Done.

pbardea commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -1249,6 +1270,52 @@ func (r *importResumer) ReportResults(ctx context.Context, resultsCh chan<- tree
+	if cfg.numIgnoredStmts < pgDumpMaxLoggedStmts {
+		numLoggedStmts = cfg.numIgnoredStmts
+	}
+	cfg.logBuffer.WriteString(fmt.Sprintf("\nLogging %d out of %d ignored stmts.\n",
```
Nit: let's avoid abbreviations in user-facing messages `s/stmts/statements`

adityamaru replied:
Done.

pbardea commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -1249,6 +1270,52 @@ func (r *importResumer) ReportResults(ctx context.Context, resultsCh chan<- tree
+	err = dest.WriteFile(ctx, pgDumpUnsupportedSchemaStmtLog,
+		bytes.NewReader(cfg.logBuffer.Bytes()))
+	if err != nil {
+		return errors.New("failed to log unsupported during IMPORT PGDUMP")
```
Should we wrap the error here, and above?

adityamaru replied:
Good point, done.

pbardea reviewed (approved):
LGTM with nits. I'm a fan of the log file approach!

miretskiy reviewed (changes requested):
> Reviewed 2 of 10 files at r8, 2 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57827)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/importccl/import_stmt.go, line 1302 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MTkosd19-O1t_lpnx3M:-MTkosd19-O1t_lpnx3N:bmj6mmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d99c8e5f260cdc7d49c3d4082887324835ead86/pkg/ccl/importccl/import_stmt.go#L1302)):*
> ```Go
> 	}
> 	cfg.logBuffer.WriteString(fmt.Sprintf("\nLogging %d out of %d ignored statements.\n",
> 		numLoggedStmts, cfg.numIgnoredStmts))
> ```

do we want to prefix every log statement with this?
___
*[pkg/ccl/importccl/import_stmt.go, line 1311 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MTkp8uE65CxQxTwfOoJ:-MTkp8uE65CxQxTwfOoK:bgeaskr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d99c8e5f260cdc7d49c3d4082887324835ead86/pkg/ccl/importccl/import_stmt.go#L1311)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
> 	dest, err := p.ExecCfg().DistSQLSrv.ExternalStorageFromURI(ctx,
> 		cfg.ignoreUnsupportedLogDest, p.User())
> 	if err != nil {
> 		return errors.Wrap(err, "failed to log unsupported stmts during IMPORT PGDUMP")
> 	}
> 	defer dest.Close()
> 	err = dest.WriteFile(ctx, pgDumpUnsupportedSchemaStmtLog,
> 		bytes.NewReader(cfg.logBuffer.Bytes()))
> ```
</details>

does it make sense to accumulate log messages (like we do), but only flush them at the end?
```
  type logger struct {}

... 
l := logger
defer l.FlushTo(dest)
l.Log()...
```


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57827)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/importccl/read_import_pgdump.go, line 52 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MTmZIa_9cZtASLtkVcR:-MTmZIaaCjO8JMuOgmS9:bvig04f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d99c8e5f260cdc7d49c3d4082887324835ead86/pkg/ccl/importccl/read_import_pgdump.go#L52)):*
> ```Go
> 	s              *bufio.Scanner
> 	copy           *postgreStreamCopy
> 	unsupportedCfg *unsupportedStmtConfig
> ```

Do we want to perhaps introduce 
___
*[pkg/ccl/importccl/read_import_pgdump.go, line 113 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/57827#-MTma3gaDz6m5yQ1qjjx:-MTma3gaDz6m5yQ1qjjy:bfr6qe3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d99c8e5f260cdc7d49c3d4082887324835ead86/pkg/ccl/importccl/read_import_pgdump.go#L113)):*
> ```Go
> 			// indicated that they want to skip these stmts during the IMPORT, then do
> 			// so here.
> 			if p.unsupportedCfg.ignoreUnsupported && errors.HasType(err, (*tree.UnsupportedError)(nil)) {
> ```

Discussed on how to better reorg the code -- basically, move this to "unsupported logger" struct and have a nice method (Log) to log.


<!-- Sent from Reviewable.io -->

adityamaru commented:
@miretskiy this should be RFAL

pbardea commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -1249,6 +1270,102 @@ func (r *importResumer) ReportResults(ctx context.Context, resultsCh chan<- tree
+
+// unsupportedStmtLogger is responsible for handling unsupported PGDUMP SQL
+// statements seen during the import.
+type unsupportedStmtLogger struct {
```
Nice!

miretskiy reviewed (approved):
> Reviewed 1 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57827)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @miretskiy, and @pbardea)



<!-- Sent from Reviewable.io -->

adityamaru commented:
TFTRs!

bors r=pbardea,miretskiy
